### DISCUSSION TITLE
Boundary Pores Labeling

### DISCUSSION THREAD

**Discussion by falvlima** [PMEAL/OpenPNM#2323] [Q&A]
Hi,

I am facing some problems while labeling the boundary pores to a network extracted using Porespy, and I would like to request some help.
Firstly I would like to check something regarding Porespy:
I have used the following code to perform the PNM generation:

# Step-by-step application of the SNOW algorithm

``` python
sigma = 3.5
dt = spim.distance_transform_edt(input=im1)
dt = spim.gaussian_filter(input=dt, sigma=sigma)
peaks = find_peaks(dt=dt)

print('Initial number of peaks: ', spim.label(peaks)[1])
peaks = trim_saddle_points(peaks=peaks, dt=dt)
print('Peaks after trimming saddle points: ', spim.label(peaks)[1])
peaks = trim_nearby_peaks(peaks=peaks, dt=dt)
peaks, N = spim.label(peaks)
print('Peaks after trimming nearby peaks: ', N)


regions = watershed(image=-dt, markers=peaks, mask=dt > 0)
regions = randomize_colors(regions)

fig, ax = plt.subplots(1,2)
ax[0].imshow((im1));
ax[1].imshow((regions*im1));


net = ps.networks.regions_to_network(regions*im1, voxel_size=1)

pn2, geo2 = op.io.PoreSpy.import_data(net, settings=settings)
```

This one does not give me labeled pores for the boundaries.

I know I could have used snow2, which would make my life easier, but could I change the sigma (or something similar in this code)? Apparently, this code brought some over-segmentation and a lack of connections between the pores. 

# Automatic Snow algorithm

``` python
resolution = 5e-5

snow = ps.networks.snow2(imbool, voxel_size=resolution)

settings = {'pore_shape': 'pyramid',
            'throat_shape': 'cuboid',
            'pore_diameter': 'equivalent_diameter',
            'throat_diameter': 'inscribed_diameter'}

pn, geo = op.io.PoreSpy.import_data(snow.network, settings=settings)
```

Second, on the network extracted using the ``ps.networks.regions_to_network()`` function, I tried to label by using the ``op.topotools.find_surface_pores`` functions. Unfortunately, when trying to extract top, bottom, left, right boundaries, I could not handle them.

![Figure_1](https://user-images.githubusercontent.com/83393087/152853598-fd826dfd-9e24-4175-800f-808bf6d149a1.png)

``` python
op.topotools.find_surface_pores(network=pn2, label='surface')
fig = op.topotools.plot_connections(network=pn2)
fig = op.topotools.plot_coordinates(network=pn2, pores=pn2.pores('surface'), ax=fig, c='r')
```

I have tried using markers in the code above (like in this example https://notebook.community/PMEAL/OpenPNM/examples/topology/Adding%20Boundary%20Pores) and setting the "top" labels afterward using the following command but without success.

``pn2.set_label(pores=pn2.pores('surface')[1:30], label='pore.top')``

Could you help me with this issue? Is there any detail regarding the markers that I am not taking into consideration?

Thanks in advance!

Fernando

**Comment by falvlima** [comment]
Ok. I have managed to do it. Apparently, the markers work like some kind of influence zone...

# Prepare the markers

``` python

numero = 100

x = np.linspace(-25, -10, num=numero)

y=np.linspace(20, 460, num=numero)

result = []

for a in x:
    for b in y:
        markers1 = np.array([a,b])
        result.append(markers1)

markers = np.array(result)

op.topotools.find_surface_pores(network=pn2, markers=markers, label='surface_left')
fig = op.topotools.plot_connections(network=pn2)
fig = op.topotools.plot_coordinates(network=pn2, pores=pn2.pores('surface_left'), ax=fig, c='r')


x, y = markers.T
plt.scatter(x,y)
plt.show()
```

![Figure_2](https://user-images.githubusercontent.com/83393087/152965340-8703f2e1-7336-439c-b4f5-01ba82f48a03.png)

Anyway, if you can help me with the "snow issue", it is still valid.

Cheers

**Reply by jgostick** [reply]
I'm not quite sure I understand where you're having a problem.  Can you clarify the specific place you're having trouble?
